File: Swift/Easy/plusOne.swift
// https://leetcode.com/problems/plus-one/

class Solution {
    func plusOne(_ digits: [Int]) -> [Int] {
        
        var dig = digits
        
        var z = dig.last!
        
        if z < 9 {
            z += 1
            dig.removeLast()
            dig.append(z)
            return dig
        }
        
        var total = 0
        var l = digits.count
        
        var b:[Int] = []
        
        for e in digits.reversed() {
            b.append(e)
        }
        
        var temp = l
        var tens = 1
        
         while temp != 1 {
                tens = tens * 10
                temp -= 1 
                }
        
        while l > 0 {
            
            let d = b.last!
            b = b.dropLast()
            
            let add = tens * d
            total = total + add
            tens = tens/10
            l -= 1
        
        }
        
        total += 1
        let s = String(total)
        let returnVal = s.map { char -> Int in 
                              if let intVal = Int(String(char)){
                                  return intVal
                              }
                                    return 0
        }
        
        return returnVal
    }
}

File: Easy/plus-one.py
''' 

https://leetcode.com/problems/plus-one/

Runtime
37 ms
Beats
40.99%
Memory
14 MB
Beats
5.53%

'''

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        digits.reverse()
        carry = 1
        add_a_place = False

        for i in range(len(digits)):
            if digits[i] != 9:
                digits[i] += carry
                carry = 0
            else:
                digits[i] += carry
                if digits[i] == 10:
                    digits [i] = 0
                    carry = 1
                    if len(digits) - 1 == i:
                        add_a_place = True

        s = set()
        for j in digits:
            s.add(j)
        
        z = set()
        z.add(0)

        if s.symmetric_difference(z) == set():
            digits.append(1)
            digits.reverse()
            return digits
        elif add_a_place == True:
            digits.append(1)
            digits.reverse()
            return digits 
        else:
            digits.reverse()
            return digits
